enhancements/network/netflow.md

```diff
@@ -0,0 +1,121 @@
+
+This proposal discusses how we start providing this functionality by allowing cluster administrators to enable the export of network flow metadata using one of these protocols.
+
+## Motivation
+
+* Cluster administrators need to be able to get network flows out of OpenShift to be consumed by their collectors.
+
+### Goals
+
+- Add support for exporting the network flows traffic in an OpenShift cluster by leveraging the NetFlow/sflow/IPFIX protocols supported by OVS.
+
+### Non-Goals
+
+- This document only applies to OVN-Kubernetes, not OpenShift SDN or any other other network plugin.
+- This document will not discuss the flow logs store solutions. Typically, customers store the NetFlow data in a data warehouse system (e.g. ClickHouse) or search and analytics system (e.g. ElasticSearch). It is out of the scope of this document to discuss which option should be used and if we should manage that solution.
+- This document will not discuss the aggregation of flow data with corresponding K8S pods.
+- This document will not discuss flow visualization or flow dashboards.
+
+## Proposal
+
+OVS supports NetFlow v5, IPFIX and sflow as export flows protocol. It also supports a list of collectors by specifying IP and port to send the flow data.
+As such, we would support the three protocols and an arbitrary list of collectors.
+
+The Cluster Network Operator (CNO) would expose:
+* *exportNetworkFlowsConfig* dict to contain the options to export the network flows
+  * *exportNetworkFlowsProtocol* to specify which protocol should be used. Valid values should be *netflow*, *ipfix* and *sflow*.
+  * *exportNetworkFlowsCollectors* list to contain the collectors IPs and ports
+    * *exportNetworkFlowsCollectorIP* to specify the IP of the collector that will consume the flow data
+    * *exportNetworkFlowsCollectorPort* to specify the port of the collector that will consume the flow data
+
+This is an example of a *networks.operator.openshift.io* spec containing two collectors:
+
+```yaml
+spec:
+  clusterNetwork:
+  - cidr: 10.128.0.0/14
+    hostPrefix: 23
+  defaultNetwork:
+    ovnKubernetesConfig:
+      exportNetworkFlowsConfig:
+        exportNetworkFlowsCollectors:
+        - exportNetworkFlowsCollectorIP: 172.30.158.150
+          exportNetworkFlowsCollectorPort: 2056
+        - exportNetworkFlowsCollectorIP: 172.30.54.103
+          exportNetworkFlowsCollectorPort: 2056
+        exportNetworkFlowsProtocol: netflow
+      genevePort: 6081
+      mtu: 1400
+    type: OVNKubernetes
+  deployKubeProxy: false
+  disableMultiNetwork: false
+  disableNetworkDiagnostics: false
+  logLevel: Normal
+  managementState: Managed
+  observedConfig: null
+  operatorLogLevel: Normal
+  serviceNetwork:
+  - 172.30.0.0/16
+```
+
+Under the covers, these options will make the CNO to perform an ovs-vsctl command on the bridge br-int, which is the bridge the containers are connected to on OVN-Kubernetes, and its management port ovn-k8s-mp0.
```

Comment:
```suggestion
Under the covers, these options will make the CNO perform an ovs-vsctl command on the bridge br-int, which is the bridge the containers are connected to on OVN-Kubernetes, and its management port ovn-k8s-mp0.
```

Reply:
Maybe not worth arguing about here but would this actually be done by CNO? I'd expect CNO to write out something to the ovn-kube config, and ovn-kube to modify `br-int`

Reply:
Why? To enable any of the export protocols you just run an ovs-vsctl command.
In my POC, I put it on the 006 -ovs-node.yaml template, seems simple enough and it's also the same pattern that ipsec feature followed.

Reply:
1. `br-int` is an internal detail of ovn-kubernetes and CNO should not be making assumptions about it. That's the sort of thing that could easily change in the future. (Or maybe local gateway and shared gateway do it differently. Or maybe some update to the external gateway handling will result in some traffic not crossing `br-int`. Etc.)
2. Non-OCP users of ovn-kubernetes may want to enable flow export too

Reply:
Oh, so you mean making this an ovn-k change:

@trozet @russellb can you please chime in to get consensus?
I always thought on doing this from CNO, per https://github.com/openshift/cluster-network-operator/pull/982 .

But if you all agree this should be upstreamed to OVN-K, then that's fine. I assume the change would be on https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/node/node.go ?

Reply:
Anyway, it's not worth discussing the implementation detail here.
Let's have that discussion over slack or PRs.